### PR TITLE
Add json format functions for key-value lists

### LINKED ISSUE #8882: json module format functions for key-value lists
Existing `json:encode_key_value_list/2` and `json:encode_key_value_list_checked/2` can be used to encode key-value lists while preserve their ordering.
 It will be great if `json` module exports corresponding
`-spec format_key_value_list([{term(), term()}], Encode::formatter(), State::map()) -> iodata().`
and
`-spec format_key_value_list_checked([{term(), term()}], Encode::formatter(), State::map()) -> iodata().`
for generating formatted JSON from key-value lists.

### PR DESCRIPTION
**Pull request by isvilen**
Support formatting key-value lists while preserve their ordering.
Closes #8882

### CODE REVIEW

File: lib/stdlib/src/json.erl [hunk 1]
```diff
@@ -694,6 +696,15 @@ format_tail([Head|Tail], Enc, State, IndentAll, IndentRow) ->
 format_tail([], _, _, _, _) ->
     [].
 
+
```
Comment (vkatsuba) [inline lib/stdlib/src/json.erl:699]:
```suggestion
```

File: lib/stdlib/src/json.erl [hunk 2]
```diff
@@ -706,6 +717,50 @@ format_key_value_list(KVList, UserEnc, #{level := Level} = State) ->
             end,
     format_object([Entry(Key,Value) || {Key, Value} <- KVList], Indent).
 
+
```
Comment (vkatsuba) [inline lib/stdlib/src/json.erl:720]:
```suggestion
```

File: lib/stdlib/src/json.erl [hunk 3]
```diff
@@ -706,6 +717,50 @@ format_key_value_list(KVList, UserEnc, #{level := Level} = State) ->
...
+                        {[EncEntry | Acc], Visited1}
+                end
+        end,
+    {EncKVList, _} = lists:foldl(EncListFun, {[], #{}}, KVList),
+    lists:reverse(EncKVList);
+
+do_format_checked([], _, _) ->
+    [].
+
+
```
Comment (vkatsuba) [inline lib/stdlib/src/json.erl:763]:
```suggestion
```

File: lib/stdlib/src/json.erl [hunk 4]
```diff
@@ -706,6 +717,50 @@ format_key_value_list(KVList, UserEnc, #{level := Level} = State) ->
...
+""".
+-doc(#{since => <<"OTP 27.2">>}).
+
+-spec format_key_value_list_checked([{term(), term()}], Encoder::formatter(), State::map()) -> iodata().
+format_key_value_list_checked(KVList, UserEnc, State) when is_function(UserEnc, 3) ->
+    {_,Indent} = indent(State),
+    format_object(do_format_checked(KVList, UserEnc, State), Indent).
+
+do_format_checked(KVList, UserEnc,  #{level := Level} = State) ->
+    NextState = State#{level := Level+1},
```
Comment (vkatsuba) [inline lib/stdlib/src/json.erl:740]:
```suggestion
    NextState = State#{level := Level + 1},
```

File: lib/stdlib/src/json.erl [hunk 5]
```diff
@@ -706,6 +716,48 @@ format_key_value_list(KVList, UserEnc, #{level := Level} = State) ->
...
+    NextState = State#{level := Level + 1},
+    {KISize, KeyIndent} = indent(NextState),
+    EncKeyFun = fun(KeyVal, _Fun) -> UserEnc(KeyVal, UserEnc, NextState) end,
+    EncListFun =
+        fun({Key, Value}, {Acc, Visited0}) ->
+                EncKey = iolist_to_binary(key(Key, EncKeyFun)),
+                case is_map_key(EncKey, Visited0) of
+                    true ->
+                        error({duplicate_key, Key});
+                    _ ->
```
Comment (Maria-12648430) [inline lib/stdlib/src/json.erl:750]:
```suggestion
                    false ->
```
Better be explicit ;)

**Review by vkatsuba — APPROVED** [review]

**Review by dgud — CHANGES_REQUESTED** [review]

**Comment by dgud** [comment]
Can you squash the commits and force-push..

**Comment by isvilen** [comment]
> Can you squash the commits and force-push..

Done, should I also rebase it onto main ?

**Comment by dgud** [comment]
I think you are already based of *maint* branch. 

I could merge it without getting any changes from master, what I saw at least, but could have missed something.
Checked and it looks ok to me, before the squash at least and still good after squash :-)

**Pull request merged**